File: VB/WinWebSolution.Module/Module.vb
Imports System
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Xpo
Imports System.Configuration

Namespace WinWebSolution.Module

    Public NotInheritable Partial Class WinWebSolutionModule
        Inherits ModuleBase

        Private Shared provider As XpoDataStoreProxyProvider

        Public Sub New()
            InitializeComponent()
        End Sub

        Public Overrides Sub Setup(ByVal application As XafApplication)
            MyBase.Setup(application)
            AddHandler application.CustomCheckCompatibility, New EventHandler(Of CustomCheckCompatibilityEventArgs)(AddressOf application_CustomCheckCompatibility)
            AddHandler application.CreateCustomObjectSpaceProvider, New EventHandler(Of CreateCustomObjectSpaceProviderEventArgs)(AddressOf application_CreateCustomObjectSpaceProvider)
        End Sub

        Private Sub application_CreateCustomObjectSpaceProvider(ByVal sender As Object, ByVal e As CreateCustomObjectSpaceProviderEventArgs)
            If provider Is Nothing Then
                provider = New XpoDataStoreProxyProvider()
            End If

            e.ObjectSpaceProvider = New XPObjectSpaceProvider(provider)
        End Sub

        Private Sub application_CustomCheckCompatibility(ByVal sender As Object, ByVal e As CustomCheckCompatibilityEventArgs)
            If provider IsNot Nothing AndAlso Not provider.IsInitialized Then
                Call provider.Initialize(CType(e.ObjectSpaceProvider, XPObjectSpaceProvider).XPDictionary, ConfigurationManager.ConnectionStrings("LegacyDatabaseConnectionString").ConnectionString, ConfigurationManager.ConnectionStrings("TempDatabaseConnectionString").ConnectionString)
            End If
        End Sub
    End Class
End Namespace

File: CS/WinWebSolution.Module/Module.cs
using System;
using System.Collections.Generic;
using DevExpress.ExpressApp;
using System.Reflection;
using DevExpress.ExpressApp.Xpo;
using System.Configuration;

namespace WinWebSolution.Module {
    public sealed partial class WinWebSolutionModule : ModuleBase {
        private static XpoDataStoreProxyProvider provider;
        public WinWebSolutionModule() {
            InitializeComponent();
        }
        public override void Setup(XafApplication application) {
            base.Setup(application);
            application.CustomCheckCompatibility += new EventHandler<CustomCheckCompatibilityEventArgs>(application_CustomCheckCompatibility);
            application.CreateCustomObjectSpaceProvider += new EventHandler<CreateCustomObjectSpaceProviderEventArgs>(application_CreateCustomObjectSpaceProvider);
        }
        void application_CreateCustomObjectSpaceProvider(object sender, CreateCustomObjectSpaceProviderEventArgs e) {
            if(provider == null) {
                provider = new XpoDataStoreProxyProvider();
            }
            e.ObjectSpaceProvider = new XPObjectSpaceProvider(provider);
        }
        void application_CustomCheckCompatibility(object sender, CustomCheckCompatibilityEventArgs e) {
            if(provider != null && !provider.IsInitialized) {
                provider.Initialize(((XPObjectSpaceProvider)e.ObjectSpaceProvider).XPDictionary,
                    ConfigurationManager.ConnectionStrings["LegacyDatabaseConnectionString"].ConnectionString,
                    ConfigurationManager.ConnectionStrings["TempDatabaseConnectionString"].ConnectionString);
            }
        }
    }
}
